Optimize compile angel  guess el file

LemonBreezes opened the pull request:
I have 10x'd the performance of this package basically by
1. let-binding `file-name-handler-alist` to `nil`. This gives the biggest performance boost.
2. Instead of searching for each package in all `load-path` directories, first scan all the `load-path` directories and create a hash table index, then use that for lookups.

There are still a lot of other significant performance gains that can be made as well.

LemonBreezes commented:
I have just made another optimization: use symbols instead of "feature names" whenever possible. This also leads to nearly a 3x speedup on my laptop over the 10x speedup I just mentioned.

LemonBreezes commented:
Another small performance boost for handling `evil-collection` specially since it loads so many packages without adding them to `load-path`.

LemonBreezes commented:
I have made it slightly faster but more importantly, I am now using `load-history` when possible for finding the file because sometimes the features we loaded are not in `load-path` directly. Also, doing so is slightly faster.

LemonBreezes commented:
I have more-judiciously added `nil` let-binds to `file-name-handler-alist` across the file. Now `compile-angel-on-load-mode` loads for me in `0.06` seconds.

LemonBreezes commented:
Now down to `0.046` seconds by disabling `case-fold-search`.

LemonBreezes commented:
By assuming the doom emacs directories are not symlinks and the files are not files that point inside the doom emacs directories (reasonable), I have optimized `compile-angel--need-compilation-p` further. Now I am getting `0.032` seconds.

jamescherti commented:
Hello @LemonBreezes,

You have implemented effective optimizations. Thank you for your valuable contribution to the Emacs community.

I look forward to evaluating your code and providing feedback.

jamescherti commented on compile-angel.el:
```diff
@@ -411,17 +449,13 @@ FEATURE-NAME is a string representing the feature name being loaded."
-                 (file-in-directory-p
-                  el-file (expand-file-name doom-modules-dir))))
+       ((and (boundp 'doom-user-dir)
+             (or (string-prefix-p doom-user-dir el-file)
```
Using  `string-prefix-p` may be faster, but `file-in-directory-p` is more accurate because it uses `file-truename` to normalize both paths (e.g., paths can differ if they are symlinks or use `~/`, path that ends with or without `/`, etc.).

I think it is better to revert to using `file-in-directory-p` for now. 

(In the future, we could optimize this by caching a version of `doom-user-dir` that has been standardized using `file-truename`.)

LemonBreezes replied:
Yep. I have just added this.

jamescherti commented on compile-angel.el:
```diff
@@ -411,17 +449,13 @@ FEATURE-NAME is a string representing the feature name being loaded."
-                  el-file (expand-file-name doom-modules-dir))))
+       ((and (boundp 'doom-user-dir)
+             (or (string-prefix-p doom-user-dir el-file)
+                 (string-prefix-p doom-emacs-dir el-file)
```
1. This needs to check if doom-emacs-dir is bound, just in case.
2. Using  `string-prefix-p` may be faster, but `file-in-directory-p` is more accurate because it uses `file-truename` to normalize both paths (e.g., paths can differ if they are symlinks or use `~/`, path that ends with or without `/`, etc.). I think it is better to revert to using `file-in-directory-p` for now.

LemonBreezes replied:
I have addressed all of those concerns by caching the truename of the expansions then using `string-prefix-p`. I have also fixed an issue where I noticed I wasn't using `~/.config/emacs/lisp`.

jamescherti commented on compile-angel.el:
```diff
@@ -411,17 +449,13 @@ FEATURE-NAME is a string representing the feature name being loaded."
+       ((and (boundp 'doom-user-dir)
+             (or (string-prefix-p doom-user-dir el-file)
+                 (string-prefix-p doom-emacs-dir el-file)
+                 (string-prefix-p doom-modules-dir el-file)))
```
1. This needs to check if doom-modules-dir is bound, just in case.
2. Using  `string-prefix-p` may be faster, but `file-in-directory-p` is more accurate because it uses `file-truename` to normalize both paths (e.g., paths can differ if they are symlinks or use `~/`, path that ends with or without `/`, etc.). I think it is better to revert to using `file-in-directory-p` for now.

LemonBreezes replied:
> 1. This should check if `doom-modules-dir` is bound to avoid making assumptions.
> 
>     2. Using  `string-prefix-p` may be faster, but `file-in-directory-p` is more accurate because it uses `file-truename` to normalize both paths (e.g., paths can differ if they are symlinks or use `~/`, path that ends with or without `/`, etc.). I think it is better to revert to using `file-in-directory-p` for now.

Those Doom directories are already expanded in the Doom core at definition time and are never symlinked by Doom (and Doom enforces not letting the user touch them!). Furthermore, they are all defined right after each other. That function is called thousands of time and any such change WILL significantly slow down the package.
![image](https://github.com/user-attachments/assets/04570230-e0b3-4f09-83b7-fed80567688d)

LemonBreezes replied:
I am writing a fix to compute the expanded filenames at the beginning when the file is loaded.

LemonBreezes replied:
Okay. I have just added a fix for this. Now I compute the truenames at package load time. I think it's a good middle ground and resolves the bytecomp warnings.

LemonBreezes commented:
Okay. I think I have made it about as fast as I can make it.

jamescherti commented on compile-angel.el:
```diff
@@ -531,58 +566,167 @@ detected, it raises an error and returns nil."
      feature (type-of feature))
     nil)))
 
+(defun compile-angel--build-file-index ()
```
I'm unsure about `compile-angel--build-file-index`, as it could introduce overhead when starting the mode, particularly when the number of `.el` files in the user's `load-path` is large.

I would prefer building such a list dynamically, before every load/require for example.

Would you mind removing this optimization. We can discuss it in another pull request.

LemonBreezes replied:
Take a look at the output of these three benchmarks (being ran after all elisp packages are loaded):
```elisp
(list
 (let ((compile-angel-use-file-index t))
   (clrhash compile-angel--list-compiled-files)
   (clrhash compile-angel--file-index)
   (benchmark-run 1
     (compile-angel-on-load-mode +1)))
 (let ((compile-angel-use-file-index nil))
   (clrhash compile-angel--list-compiled-files)
   (clrhash compile-angel--file-index)
   (benchmark-run 1
     (compile-angel-on-load-mode +1)))
 (progn
   (clrhash compile-angel--list-compiled-files)
   (clrhash compile-angel--file-index)
   (benchmark-run 1
     (compile-angel--build-file-index))))
;; => ((0.050082265 0 0.0) (2.2902459470000003 0 0.0) (0.018014330000000002 0 0.0))
```
And this is on an ultra-thin laptop from 2018.

It is VASTLY faster to create the index then use the index to find an elisp feature than to find all the elisp features one-by-one as Emacs is loading. When Emacs is loading, This is because we are searching through the `load-path` in one pass rather than searching through it for each individual package. Furthermore, when packages are already loaded, we can search for them in `load-history` instead which is much faster AND more accurate than searching through `load-path`.

LemonBreezes replied:
Also, Doom Emacs does not give us a way to load elisp before Doom Emacs's core is loaded so we can't assume `compile-angel` will always be ran before everything else. In fact, it's better to actually run `compile-angel` AFTER everything else is loaded.

jamescherti commented on compile-angel.el:
```diff
@@ -660,41 +810,35 @@ existence.
-     "compile-angel--find-el-file: NO .el FILE CORRESPONDS TO: %s" file))))
+  (compile-angel--with-fast-file-ops
+    (cond
+     ((not file) nil)
```
Please add back:
```
((not file)
    (compile-angel--debug-message
     "compile-angel--find-el-file: nil file")
    nil)
```

It can be useful for debugging purposes.

LemonBreezes replied:
Okay. I have re-added it back in.

jamescherti commented on compile-angel.el:
```diff
@@ -660,41 +810,35 @@ existence.
+      (let* ((base (file-name-sans-extension file))
+             (plain-el (concat base ".el")))
+        ;; First try plain .el as it's most common
+        (if (file-exists-p plain-el)
```
This will result in `file-exists-p` being called twice if the `.el` file does not exist. (e.g., If the file is an .el.gz for example). It is better, for now, to just use:
```elisp
 (cl-some (lambda (suffix)
                     (let ((candidate (concat base ".el" suffix)))
                       (and (file-exists-p candidate) candidate)))
                   load-file-rep-suffixes)))
```

LemonBreezes replied:
Okay. Just fixed this.

jamescherti reviewed (changes requested):
Here are my comments.

jamescherti commented:
In addition to my previous thoughts, I tested the code and it appears to fail because it attempts to compile a directory:
![image](https://github.com/user-attachments/assets/1159d4b4-39a3-4cf3-8586-bea375002b79)

jamescherti reviewed (dismissed):
This recent change results in a byte-compilation issue:
```
cask emacs -batch -L . \
          --eval "(setq byte-compile-error-on-warn t)" \
  -f batch-byte-compile $(cask files); \
  (ret=$? ; cask clean-elc && exit $ret)

In toplevel form:
compile-angel.el:455:35:Error: reference to free variable ‘doom-user-dir’
make: *** [Makefile:42: compile] Error 1
```

LemonBreezes commented:
> I tested the code and it appears to fail because it attempts to compile a directory: ![image](https://private-user-images.githubusercontent.com/60946298/423130704-1159d4b4-39a3-4cf3-8586-bea375002b79.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIwNTc2NDMsIm5iZiI6MTc0MjA1NzM0MywicGF0aCI6Ii82MDk0NjI5OC80MjMxMzA3MDQtMTE1OWQ0YjQtMzlhMy00Y2YzLTg1ODYtYmVhMzc1MDAyYjc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE1VDE2NDkwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxNTExMTg4N2U0ZTEyZWI2NDRmN2MwM2FmODg0YWIxM2MwNDE1MGY1NjAyZGM3YzlmNTEwYmU2MDc4ZWJiNTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3ofbluMnlD7k-cMq0in6tudH3QJZ3gi603O4i8kUsok)

Try it again. I have just made a lot of changes and addressed all the points you mentioned. Let me know if you still get that error.

jamescherti commented:
I tried again. Same issue:
![image](https://github.com/user-attachments/assets/3514feb7-3f17-463a-a44d-bfde3cdb0a50)

There is also also a `compile-angel.el:262:41:Error: reference to free variable ‘doom-user-dir’`:
```elisp
Loading package information... done
Package operations: 1 installs, 0 removals
  - Installing [ 1/1] emacs (26.3)... already present
cask emacs -batch -L . \
          --eval "(setq byte-compile-error-on-warn t)" \
  -f batch-byte-compile $(cask files); \
  (ret=$? ; cask clean-elc && exit $ret)

In toplevel form:
compile-angel.el:262:41:Error: reference to free variable ‘doom-user-dir’
make: *** [Makefile:42: compile] Error 1
```

This pull request includes too many changes, making it difficult to review effectively.

It might be better to remove the features that have issues, such as `compile-angel-use-file-index` (including the report), and submit them in a separate pull request once this one is merged.

How does that sound?

jamescherti commented:
Another solution would be to disable the file index by default for now, until all bugs are tested and fixed.

jamescherti commented:
```
cask emacs -batch -L . \
          --eval "(setq byte-compile-error-on-warn t)" \
  -f batch-byte-compile $(cask files); \
  (ret=$? ; cask clean-elc && exit $ret)

In toplevel form:
compile-angel.el:817:1:Error: Unused lexical variable ‘plain-el’
make: *** [Makefile:42: compile] Error 1
```
I recommend using https://github.com/purcell/package-lint to automatically detect these issues with Flymake or Flycheck.

LemonBreezes commented:
> I tried again. Same issue: ![image](https://private-user-images.githubusercontent.com/60946298/423158622-3514feb7-3f17-463a-a44d-bfde3cdb0a50.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIwNjk1NTAsIm5iZiI6MTc0MjA2OTI1MCwicGF0aCI6Ii82MDk0NjI5OC80MjMxNTg2MjItMzUxNGZlYjctM2YxNy00NjNhLWE0NGQtYmZkZTNjZGIwYTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE1VDIwMDczMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQxNDg4MjUzZGVhMjQ0MTliNTAzZjRmYzIyOTI4NmZjMTY1NTE4ZmM4NjU3NWVjNGIyYTcxZjYxZGYwYzNlMDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MctwdSQClmqe2nA8S-_T6MnDYX5cuiW2hA093CGb3Q8)
> 
> There is also also a `compile-angel.el:262:41:Error: reference to free variable ‘doom-user-dir’`:
> 
> ```emacs-lisp
> Loading package information... done
> Package operations: 1 installs, 0 removals
>   - Installing [ 1/1] emacs (26.3)... already present
> cask emacs -batch -L . \
>           --eval "(setq byte-compile-error-on-warn t)" \
>   -f batch-byte-compile $(cask files); \
>   (ret=$? ; cask clean-elc && exit $ret)
> 
> In toplevel form:
> compile-angel.el:262:41:Error: reference to free variable ‘doom-user-dir’
> make: *** [Makefile:42: compile] Error 1
> ```
> 
> This pull request includes too many changes, making it difficult to review effectively.
> 
> It might be better to remove the features that have issues, such as `compile-angel-use-file-index` (including the report), and submit them in a separate pull request once this one is merged.
> 
> How does that sound?

That warning is just a warning, not an actual error. I have now silenced the byte compiler again. Try the code as it is. It works.

LemonBreezes commented:
Just give it one try. Trust me. Run the benchmark:
```elisp
(list
 (let ((compile-angel-use-file-index t))
   (clrhash compile-angel--list-compiled-files)
   (clrhash compile-angel--file-index)
   (benchmark-run 1
     (compile-angel-on-load-mode +1)))
 (let ((compile-angel-use-file-index nil))
   (clrhash compile-angel--list-compiled-files)
   (clrhash compile-angel--file-index)
   (benchmark-run 1
     (compile-angel-on-load-mode +1)))
 (progn
   (clrhash compile-angel--list-compiled-files)
   (clrhash compile-angel--file-index)
   (benchmark-run 1
     (compile-angel--build-file-index))))
```

jamescherti commented:
It can't test the benchmark because of the bug I mentioned (it tries to compile a directory for some reason, possibly because the directory contains `.el` in its name).

I'll merge this pull request in a few hours and make some adjustments myself in the develop branch. Feel free to send any final changes before the code is merged.

Thank you again for your contribution., @LemonBreezes.

jamescherti commented:
I updated and merged this pull request, temporarily removing `compile-angel-use-file-index` (I'll move its code to a separate branch and test it later.)

jamescherti commented:
Hello @LemonBreezes,

I have merged your contribution into the `develop` branch. Could you test it and confirm that it works with your configuration? Thanks.

LemonBreezes commented:
> Hello @LemonBreezes,
> 
> I have merged your contribution into the `develop` branch. Could you test it and confirm that it works with your configuration? Thanks.

Hello. I am not getting any errors but I noticed that because you removed the cond statement from `compile-angel--guess-el-file`, the statements no longer short-circuit when we find the file, but rather we trigger ALL of the statements. For there to be any performance benefit from the file index or using the `load-history`, we need to let the statement short-circuit, either with a `cond` or a `pcase` or a `cl-return`, etc.

jamescherti commented:
In this new version, the function uses `result` to check for a value, allowing other statements to be skipped when `result` is non-nil.

LemonBreezes commented:
> Thanks for confirming that it works.
> 
> In this version, the function uses `result` to check for a value, skipping other statements when `result` is non-nil. I may revise it in the future.

Huh, it is still not short-circuiting. `compile-angel--guess-el-file` is trying every single heuristic, even if the one before already guessed the file location.

LemonBreezes commented:
The new version is completely broken for me also because it is following the symlinks for the .el files which is NOT INTENTIONAL for Doom Emacs!

jamescherti commented:
> I am expecting 0.05s for building the file index AND searching the entire load-history on my Emacs. What I am seeing is that the times roughly are 0.05s HIGHER in the cache version, indicating that we are finding the file in the cache and running the search anyways.

I'll address this issue later.

> The new version is completely broken for me also because it is following the symlinks for the .el files which is NOT INTENTIONAL for Doom Emacs! We should be building the .elc files in the directory of the symlink, not the truename for the file. Emacs Straight uses symlinks in such a way that we DO NOT want to follow them.

Thank you for pointing that out. I have removed `file-truename`, which I originally added to make `string-prefix-p` work with Doom.  

Does the latest version in the `develop` branch resolve the issue?

jamescherti commented:
Unfortunately, after reintegrating the file index function, I tested it again, and it is still failing:
![image](https://github.com/user-attachments/assets/bf81a698-9b09-429b-95a7-7b9d69808320)

LemonBreezes commented:
> Unfortunately, after reintegrating the file index function, I tested it again, and it is still failing: ![image](https://private-user-images.githubusercontent.com/60946298/424157252-bf81a698-9b09-429b-95a7-7b9d69808320.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzMzQ0MjAsIm5iZiI6MTc0MjMzNDEyMCwicGF0aCI6Ii82MDk0NjI5OC80MjQxNTcyNTItYmY4MWE2OTgtOWIwOS00MjliLTk1YTctN2I5ZDY5ODA4MzIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE4VDIxNDIwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFkNjI2OGYxMjg5YjMyODI2ZGNmNWUzNjM1YjlmYTU0Mzc1M2IxZjY3MDM3NDYwNWY1ZDE5MDhjOWExZDFlNmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.imS-NNWv3klLYRuABnmQ2SOpfttVidrHPEMIV3Fwyk8)

Okay. I am pretty sure I know what is happening. For one thing, if you run this code:
```elisp
(maphash (lambda (key value)
           (when (file-directory-p value)
             (message "key: %s, value: %s" key value)))
         compile-angel--file-index)
```
you probably see a few directories getting spit out. The correct fix is to add a `file-regular-p` check to `compile-angel--build-file-index`
```elisp
    ;; Pre-compute the combined pattern once
    (let* ((combined-pattern (concat "\\`[^.].*\\("
                                     (mapconcat #'regexp-quote
                                                compile-angel--el-file-extensions
                                                "\\|")
                                     "\\)\\'"))
           ;; Filter load-path once
           (filtered-load-path (cl-remove-if-not #'file-directory-p load-path)))

      ;; Process each directory in load-path
      (dolist (dir filtered-load-path)
        (dolist (file (directory-files dir t combined-pattern t))
          ;; Extract the feature symbol from the filename - intern directly
          (when (file-regular-p file)
            (let ((feature-symbol (intern (file-name-base file))))
              ;; Store in the index, with the first occurrence taking precedence
              (unless (gethash feature-symbol compile-angel--file-index)
                (puthash feature-symbol file compile-angel--file-index)))))))
```
Otherwise, it will add directories like "compile-angel.el" to the hash table.

LemonBreezes commented:
Don't worry. I'll work through all the bugs this change causes. I think this is a really really good idea for a package and we are so close to that sweet performance gain at a tiny fraction of a second cost.

jamescherti commented:
> you probably see a few directories getting spit out. The correct fix is to add a file-regular-p check to compile-angel--build-file-index

Check the latest commit on the `develop` branch. Unfortunately, there is still an issue, even after verifying that the file exists (though it's a different issue):
![image](https://github.com/user-attachments/assets/711b07e1-15a3-425a-a0a6-77118d5b17d2)


> Don't worry. I'll work through all the bugs this change causes. I think this is a really really good idea for a package and we are so close to that sweet performance gain at a tiny fraction of a second cost.

Sounds good.

Please make sure to base your work on the latest version of the `develop` branch to avoid merge issues.

LemonBreezes commented:
> > you probably see a few directories getting spit out. The correct fix is to add a file-regular-p check to compile-angel--build-file-index
> 
> Check the latest commit on the `develop` branch. Unfortunately, there is still an issue, even after verifying that the file exists (though it's a different issue): ![image](https://private-user-images.githubusercontent.com/60946298/424216922-711b07e1-15a3-425a-a0a6-77118d5b17d2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzMzc1NTUsIm5iZiI6MTc0MjMzNzI1NSwicGF0aCI6Ii82MDk0NjI5OC80MjQyMTY5MjItNzExYjA3ZTEtMTVhMy00MjVhLWEwYTYtNzcxMThkNWIxN2QyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE4VDIyMzQxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhYjVhNGYyNDQ1NWY5ZWFjODExZGEzZDczNDMwYmFkZThlY2Y4YzYyY2RjZjE3NjVlNDNiYTJmZWZkMTAzNjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iWrknExiTxQviNd6RcoOXTidj2wMu69m5CEHnrg6VXo)
> 
> > Don't worry. I'll work through all the bugs this change causes. I think this is a really really good idea for a package and we are so close to that sweet performance gain at a tiny fraction of a second cost.
> 
> Sounds good.
> 
> Please make sure to base your work on the latest version of the `develop` branch to avoid merge issues.

The problem is that this commit (https://github.com/LemonBreezes/compile-angel.el/commit/a66bd79b0436aa5cffc635bbb2d0a6d9f4b83b0a) changed the call to `compile-angel--locate-feature-file` to use a symbol instead of a string when it is searching for a file name.
![image](https://github.com/user-attachments/assets/166c98d2-660d-4288-b654-1cd40ba058f1)

jamescherti commented:
> The problem is that this commit (https://github.com/LemonBreezes/compile-angel.el/commit/a66bd79b0436aa5cffc635bbb2d0a6d9f4b83b0a) changed the call to compile-angel--locate-feature-file to use a symbol instead of a string when it is searching for a file name.

I have updated the argument:: 016b003d60d7870f2bf8d45dac4ed72019661892

LemonBreezes commented:
> > The problem is that this commit ([LemonBreezes@a66bd79](https://github.com/LemonBreezes/compile-angel.el/commit/a66bd79b0436aa5cffc635bbb2d0a6d9f4b83b0a)) changed the call to compile-angel--locate-feature-file to use a symbol instead of a string when it is searching for a file name.
> 
> I have updated the `compile-angel--locate-feature-file` argument:: [016b003](https://github.com/jamescherti/compile-angel.el/commit/016b003d60d7870f2bf8d45dac4ed72019661892)

Okay, so there have been some regressions between the version I submitted and the changes you made to it. The performance is still an order of magnitude worse on this new version and I have not been able to figure out why. There are also some new regressions with files being native compiled even though I have that disabled.

LemonBreezes commented:
I believe there is a problem with the implementation of `compile-angel--track-no-byte-compile-files` or one of those 3 related hash tables. It seems like there is a mess in that code because we are basically using 3 separate hash tables for the same "thing".

LemonBreezes commented:
I am going to step back and use my code for a bit since it seems the only problem with the code I wrote was the missing `file-regular-p` so I will be using that https://github.com/LemonBreezes/compile-angel.el/tree/file-index-optimization. It's faster and does not have the new issues.
